### ISSUE TITLE
Openbabel + ScikitLearn Cause Segfaults

### ISSUE THREAD

**Issue by rbharath** [deepchem/deepchem#329]
Our conda installation of openbabel and scikit-learn cause segfaults when imported together. The following simple script segfaults

```
import sklearn
import sklearn.datasets
import numpy as np
import openbabel as ob
from sklearn.linear_model import LogisticRegression

dataset = sklearn.datasets.load_digits(n_class=2)
X, y = dataset.data, dataset.target

sklearn_model = LogisticRegression()
sklearn_model.fit(X, y)
```

while the following script (only difference is that import of `openbabel` is removed) works fine:

```
import sklearn
import sklearn.datasets
import numpy as np
from sklearn.linear_model import LogisticRegression

dataset = sklearn.datasets.load_digits(n_class=2)
X, y = dataset.data, dataset.target

sklearn_model = LogisticRegression()
sklearn_model.fit(X, y)
```

Need to figure out what's going wrong here...

**Comment by rbharath** [comment]
Discovered in #327

**Comment by proteneer** [comment]
scikit-learn==0.18.1
openbabel==2.4.0

``` python
from sklearn.linear_model import LogisticRegression
import openbabel as ob
import sklearn.datasets

dataset = sklearn.datasets.load_digits(n_class=2)
X, y = dataset.data, dataset.target

sklearn_model = LogisticRegression()
sklearn_model.fit(X, y)
```

importing openbabel after linear_model fixes the problem due to presumably a shared re-initialization condition in libsvm:

```
Thread 1 "python" received signal SIGSEGV, Segmentation fault.
set_parameter (param=0x0, svm_type=2, kernel_type=24540128, degree=26063616, gamma=0.0001, coef0=1, nu=0.10000000000000001, cache_size=0, 
    C=1.3852388523421298e-309, eps=0, p=0.0001, shrinking=100, probability=775647528, nr_weight=-581718640, 
    weight_label=0x7fffdcc804e0 "\002", weight=0x7fffdcc804e0 "\002", max_iter=-590871568, random_seed=0)
    at sklearn/svm/src/libsvm/libsvm_helper.c:63
63	sklearn/svm/src/libsvm/libsvm_helper.c: No such file or directory.
(gdb) bt
#0  set_parameter (param=0x0, svm_type=2, kernel_type=24540128, degree=26063616, gamma=0.0001, coef0=1, nu=0.10000000000000001, 
    cache_size=0, C=1.3852388523421298e-309, eps=0, p=0.0001, shrinking=100, probability=775647528, nr_weight=-581718640, 
    weight_label=0x7fffdcc804e0 "\002", weight=0x7fffdcc804e0 "\002", max_iter=-590871568, random_seed=0)
    at sklearn/svm/src/libsvm/libsvm_helper.c:63
#1  0x00007fffddf75c9c in __pyx_pf_7sklearn_3svm_9liblinear_train_wrap (__pyx_v_X=__pyx_v_X@entry=0x7fffdcc80210, 
    __pyx_v_Y=__pyx_v_Y@entry=0x7fffdcc80530, __pyx_v_is_sparse=__pyx_v_is_sparse@entry=0, 
    __pyx_v_solver_type=__pyx_v_solver_type@entry=0, __pyx_v_eps=__pyx_v_eps@entry=0.0001, __pyx_v_bias=__pyx_v_bias@entry=1, __pyx_v_C=1, 
    __pyx_v_class_weight=__pyx_v_class_weight@entry=0x7fffdcc803f0, __pyx_v_max_iter=100, __pyx_v_random_seed=775647528, 
    __pyx_v_epsilon=0.10000000000000001, __pyx_v_sample_weight=0x7fffdcc80580, __pyx_self=<optimized out>) at sklearn/svm/liblinear.c:1984
#2  0x00007fffddf791cd in __pyx_pw_7sklearn_3svm_9liblinear_1train_wrap (__pyx_self=<optimized out>, __pyx_args=<optimized out>, 
    __pyx_kwds=<optimized out>) at sklearn/svm/liblinear.c:1716
```

Note that set_parameter is passed in a null pointer (param=0x0), which then gets dereferenced immediately into oblivion:

```
void set_parameter(struct svm_parameter *param, int svm_type, int kernel_type, int degree,
		double gamma, double coef0, double nu, double cache_size, double C,
		double eps, double p, int shrinking, int probability, int nr_weight,
		char *weight_label, char *weight, int max_iter, int random_seed)
{
    param->svm_type = svm_type;
...
}
```

As to why in god's name a linear solver is invoking SVM methods is beyond me, I'd need to build sklearn from source and look at the cython-generated .c files to figure that out. I can dig into this deeper if it's a blocker.

**Comment by rbharath** [comment]
Good detective work! :+1: 

No worries about digging deeper. We're trying to phase out openbabel usage over the next few weeks, so we can probably just let this one lie for the time being.

**Comment by evanfeinberg** [comment]
Thanks for figuring this out @proteneer. As much as ```openbabel``` is a
powerful tool that we've used to great effect, especially in
```grid_featurizer```, there are issues with its distribution so it will
ultimately likely have to be phased out anyway as @rbharath says above.
While I've used it much less overall, ```rdkit``` really seems like a gem
and will probably become the workhorse of ```deepchem``` for molecule
manipulation needs.

On Tue, Feb 14, 2017 at 1:03 PM, Bharath Ramsundar <notifications@github.com
